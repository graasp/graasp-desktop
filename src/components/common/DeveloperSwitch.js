import React, { Component } from 'react';
import FormControl from '@material-ui/core/FormControl';
import { withStyles } from '@material-ui/core/styles';
import { withTranslation } from 'react-i18next';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Switch from '@material-ui/core/Switch';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import { getDeveloperMode, setDeveloperMode } from '../../actions';
import Loader from './Loader';

const styles = theme => ({
  formControl: {
    margin: theme.spacing(),
    minWidth: 120,
  },
});

export class DeveloperSwitch extends Component {
  static propTypes = {
    developerMode: PropTypes.bool.isRequired,
    activity: PropTypes.bool.isRequired,
    t: PropTypes.func.isRequired,
    dispatchGetDeveloperMode: PropTypes.func.isRequired,
    dispatchSetDeveloperMode: PropTypes.func.isRequired,
    classes: PropTypes.shape({
      formControl: PropTypes.string.isRequired,
    }).isRequired,
  };

  constructor(props) {
    super(props);
    const { dispatchGetDeveloperMode } = this.props;
    dispatchGetDeveloperMode();
  }

  handleChange = async ({ target }) => {
    const { dispatchSetDeveloperMode } = this.props;
    const { checked } = target;
    dispatchSetDeveloperMode(checked);
  };

  render() {
    const { classes, t, developerMode, activity } = this.props;

    if (activity) {
      return <Loader />;
    }

    const control = (
      <Switch
        checked={developerMode}
        onChange={this.handleChange}
        value={developerMode}
        color="primary"
      />
    );

    return (
      <FormControl className={classes.formControl}>
        <FormControlLabel control={control} label={t('Developer Mode')} />
      </FormControl>
    );
  }
}

const mapStateToProps = ({ User }) => ({
  developerMode: User.getIn(['current', 'developerMode']),
  activity: Boolean(User.getIn(['current', 'activity']).size),
});

const mapDispatchToProps = {
  dispatchGetDeveloperMode: getDeveloperMode,
  dispatchSetDeveloperMode: setDeveloperMode,
};

const ConnectedComponent = connect(
  mapStateToProps,
  mapDispatchToProps
)(DeveloperSwitch);

const StyledComponent = withStyles(styles)(ConnectedComponent);

const TranslatedComponent = withTranslation()(StyledComponent);

export default TranslatedComponent;
