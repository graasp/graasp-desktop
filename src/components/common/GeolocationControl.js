import React, { Component } from 'react';
import FormControl from '@material-ui/core/FormControl';
import { withStyles } from '@material-ui/core/styles';
import { withTranslation } from 'react-i18next';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Switch from '@material-ui/core/Switch';
import Button from '@material-ui/core/Button';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import { getGeolocationEnabled, setGeolocationEnabled } from '../../actions';
import Loader from './Loader';
import { CONTROL_TYPES } from '../../config/constants';

const styles = theme => ({
  formControl: {
    margin: theme.spacing(),
    minWidth: 120,
  },
});

export class GeolocationControl extends Component {
  static propTypes = {
    geolocationEnabled: PropTypes.bool.isRequired,
    activity: PropTypes.bool.isRequired,
    t: PropTypes.func.isRequired,
    dispatchGetGeolocationEnabled: PropTypes.func.isRequired,
    dispatchSetGeolocationEnabled: PropTypes.func.isRequired,
    classes: PropTypes.shape({
      formControl: PropTypes.string.isRequired,
      button: PropTypes.string.isRequired,
    }).isRequired,
    controlType: PropTypes.oneOf(Object.keys(CONTROL_TYPES)),
  };

  static defaultProps = {
    controlType: CONTROL_TYPES.SWITCH,
  };

  constructor(props) {
    super(props);
    const { dispatchGetGeolocationEnabled } = this.props;
    dispatchGetGeolocationEnabled();
  }

  handleClick = async () => {
    const { dispatchSetGeolocationEnabled } = this.props;
    dispatchSetGeolocationEnabled(true);
  };

  handleChange = async ({ target }) => {
    const { dispatchSetGeolocationEnabled } = this.props;
    const { checked } = target;
    dispatchSetGeolocationEnabled(checked);
  };

  render() {
    const {
      classes,
      t,
      geolocationEnabled,
      activity,
      controlType = CONTROL_TYPES.SWITCH,
    } = this.props;

    if (activity) {
      return <Loader />;
    }

    const switchControl = (
      <Switch
        checked={geolocationEnabled}
        onChange={this.handleChange}
        value={geolocationEnabled}
        color="primary"
      />
    );

    return (
      <FormControl className={classes.formControl}>
        {(() => {
          switch (controlType) {
            case CONTROL_TYPES.BUTTON:
              return (
                <Button
                  variant="contained"
                  onClick={this.handleClick}
                  color="primary"
                  className={classes.button}
                >
                  {t('Enable Geolocation')}
                </Button>
              );
            case CONTROL_TYPES.SWITCH:
            default:
              return (
                <FormControlLabel
                  control={switchControl}
                  label={t('Geolocation Enabled')}
                />
              );
          }
        })()}
      </FormControl>
    );
  }
}

const mapStateToProps = ({ User }) => ({
  geolocationEnabled: User.getIn(['current', 'geolocationEnabled']),
  activity: Boolean(User.getIn(['current', 'activity']).size),
});

const mapDispatchToProps = {
  dispatchGetGeolocationEnabled: getGeolocationEnabled,
  dispatchSetGeolocationEnabled: setGeolocationEnabled,
};

const ConnectedComponent = connect(
  mapStateToProps,
  mapDispatchToProps
)(GeolocationControl);

const StyledComponent = withStyles(styles)(ConnectedComponent);

const TranslatedComponent = withTranslation()(StyledComponent);

export default TranslatedComponent;
