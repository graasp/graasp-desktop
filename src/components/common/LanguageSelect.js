import React, { Component } from 'react';
import FormControl from '@material-ui/core/FormControl';
import InputLabel from '@material-ui/core/InputLabel';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import { withStyles } from '@material-ui/core/styles';
import { withTranslation } from 'react-i18next';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { langs } from '../../config/i18n';
import { getLanguage, setLanguage } from '../../actions';
import Loader from './Loader';

const styles = theme => ({
  formControl: {
    margin: theme.spacing(),
    minWidth: 120,
  },
});

export class LanguageSelect extends Component {
  static propTypes = {
    activity: PropTypes.bool.isRequired,
    i18n: PropTypes.shape({
      changeLanguage: PropTypes.func.isRequired,
    }).isRequired,
    lang: PropTypes.string.isRequired,
    t: PropTypes.func.isRequired,
    dispatchSetLanguage: PropTypes.func.isRequired,
    dispatchGetLanguage: PropTypes.func.isRequired,
    classes: PropTypes.shape({
      formControl: PropTypes.string.isRequired,
    }).isRequired,
  };

  constructor(props) {
    super(props);
    const { dispatchGetLanguage } = this.props;
    dispatchGetLanguage();
  }

  handleChangeLanguage = async ({ target }) => {
    const { i18n, dispatchSetLanguage } = this.props;
    const { value: lang } = target;
    await i18n.changeLanguage(lang);
    dispatchSetLanguage({ lang });
  };

  renderLanguageOptions = () =>
    Object.keys(langs).map(lang => (
      <MenuItem value={lang} key={lang}>
        <em>{langs[lang]}</em>
      </MenuItem>
    ));

  render() {
    const { classes, t, lang, activity } = this.props;

    if (activity) {
      return <Loader />;
    }

    return (
      <FormControl className={classes.formControl}>
        <InputLabel htmlFor="lang">{t('Language')}</InputLabel>
        <Select
          value={lang}
          onChange={this.handleChangeLanguage}
          inputProps={{
            name: 'lang',
            id: 'lang',
          }}
        >
          {this.renderLanguageOptions()}
        </Select>
      </FormControl>
    );
  }
}

const mapStateToProps = ({ User }) => ({
  lang: User.getIn(['current', 'lang']),
  activity: Boolean(User.getIn(['current', 'activity']).size),
});

const mapDispatchToProps = {
  dispatchGetLanguage: getLanguage,
  dispatchSetLanguage: setLanguage,
};

const ConnectedComponent = connect(
  mapStateToProps,
  mapDispatchToProps
)(LanguageSelect);

const StyledComponent = withStyles(styles)(ConnectedComponent);

const TranslatedComponent = withTranslation()(StyledComponent);

export default TranslatedComponent;
