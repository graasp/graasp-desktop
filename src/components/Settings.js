import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router';
import classNames from 'classnames';
import { withStyles } from '@material-ui/core/styles';
import Drawer from '@material-ui/core/Drawer';
import CssBaseline from '@material-ui/core/CssBaseline';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import Divider from '@material-ui/core/Divider';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft';
import ChevronRightIcon from '@material-ui/icons/ChevronRight';
import { FormGroup } from '@material-ui/core';
import { withTranslation } from 'react-i18next';
import Styles from '../Styles';
import MainMenu from './common/MainMenu';
import LanguageSelect from './common/LanguageSelect';
import DeveloperSwitch from './common/DeveloperSwitch';
import GeolocationControl from './common/GeolocationControl';

class Settings extends Component {
  state = {
    open: false,
  };

  static propTypes = {
    t: PropTypes.func.isRequired,
    classes: PropTypes.shape({}).isRequired,
    theme: PropTypes.shape({}).isRequired,
    history: PropTypes.shape({
      replace: PropTypes.func.isRequired,
    }).isRequired,
    i18n: PropTypes.shape({
      changeLanguage: PropTypes.func.isRequired,
    }).isRequired,
  };

  handleDrawerOpen = () => {
    this.setState({ open: true });
  };

  handleDrawerClose = () => {
    this.setState({ open: false });
  };

  render() {
    const { classes, theme, t } = this.props;
    const { open } = this.state;

    return (
      <div className={classes.root}>
        <CssBaseline />
        <AppBar
          position="fixed"
          className={classNames(classes.appBar, {
            [classes.appBarShift]: open,
          })}
        >
          <Toolbar disableGutters={!open}>
            <IconButton
              color="inherit"
              aria-label="Open drawer"
              onClick={this.handleDrawerOpen}
              className={classNames(classes.menuButton, open && classes.hide)}
            >
              <MenuIcon />
            </IconButton>
          </Toolbar>
        </AppBar>
        <Drawer
          className={classes.drawer}
          variant="persistent"
          anchor="left"
          open={open}
          classes={{
            paper: classes.drawerPaper,
          }}
        >
          <div className={classes.drawerHeader}>
            <IconButton onClick={this.handleDrawerClose}>
              {theme.direction === 'ltr' ? (
                <ChevronLeftIcon />
              ) : (
                <ChevronRightIcon />
              )}
            </IconButton>
          </div>
          <Divider />
          <MainMenu />
        </Drawer>
        <main
          className={classNames(classes.content, {
            [classes.contentShift]: open,
          })}
        >
          <div className={classes.drawerHeader} />
          <div className={classes.settings}>
            <Typography variant="h5" color="inherit">
              {t('Settings')}
            </Typography>
            <FormGroup>
              <LanguageSelect />
              <DeveloperSwitch />
              <GeolocationControl />
            </FormGroup>
          </div>
        </main>
      </div>
    );
  }
}

const StyledComponent = withStyles(Styles, { withTheme: true })(Settings);

const TranslatedComponent = withTranslation()(StyledComponent);

export default withRouter(TranslatedComponent);
