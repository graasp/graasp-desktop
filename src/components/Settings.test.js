import React from 'react';
import { shallow } from 'enzyme';
import Typography from '@material-ui/core/Typography';
import { Settings } from './Settings';
import LanguageSelect from './common/LanguageSelect';
import DeveloperSwitch from './common/DeveloperSwitch';
import GeolocationControl from './common/GeolocationControl';

const createSettingsProps = () => {
  return {
    t: text => text,
    classes: {
      appBar: '',
      root: '',
      appBarShift: '',
      menuButton: '',
      hide: '',
      drawer: '',
      drawerPaper: '',
      drawerHeader: '',
      content: '',
      contentShift: '',
      settings: '',
    },
    theme: {
      direction: 'ltr',
    },
    history: {
      replace: jest.fn(),
    },
    i18n: {
      changeLanguage: jest.fn(),
    },
  };
};

describe('<Settings />', () => {
  let wrapper;

  beforeAll(() => {
    const props = createSettingsProps();
    // eslint-disable-next-line react/jsx-props-no-spreading
    wrapper = shallow(<Settings {...props} />);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders one <Typography /> component with text Settings', () => {
    const typography = wrapper.find(Typography);
    expect(typography).toHaveLength(1);
    expect(typography.contains('Settings')).toBeTruthy();
  });

  it('renders one <LanguageSelect /> component', () => {
    expect(wrapper.find(LanguageSelect)).toHaveLength(1);
  });

  it('renders one <DeveloperSwitch /> component', () => {
    expect(wrapper.find(DeveloperSwitch)).toHaveLength(1);
  });

  it('renders one <GeolocationControl /> component', () => {
    expect(wrapper.find(GeolocationControl)).toHaveLength(1);
  });
});
