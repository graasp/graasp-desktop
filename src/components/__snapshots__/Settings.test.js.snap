// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> renders correctly 1`] = `ShallowWrapper {}`;
